### PR TITLE
DAG

### PR DESCRIPTION
**Pull request by ejholmes**
Third times the charm :)

Fixes https://github.com/remind101/stacker/issues/300
Fixes https://github.com/remind101/stacker/issues/186
Fixes https://github.com/remind101/stacker/issues/384

This is an internal refactor to change the Plan implementation to be backed by a Directed Acyclic Graph. Most of this work was already done a long time ago in https://github.com/remind101/stacker/pull/406, but this just brings it up to speed with the master branch, with a few important caveats:

### Not parallelized

This implementation is _not_ parallelized. This PR paves the way to make it easy to implement a 100% parallelized graph walk, but I didn't implement it here, primarily to make review easier, but also because it doesn't matter much based on the current implementation (more on that below). I think we should get the graph in `master`, and then add a parallelized walk after the fact since that will need special attention to deal with concurrency issues (logging, user input, etc).

In theory, this sounds like it should reduce performance, but ironically it has little impact. I benchmarked before and after on our internal stacker graph, which is pretty big:

**before (stacker 1.1.3)**

```console
$ time stacker build

real    8m31.716s
user    0m0.139s
sys     0m0.100s
```

**after (dag)**

```console
$ time stacker build

real    8m33.819s
user    0m0.142s
sys     0m0.094s
```

However, because the `--stacks` flag now works like you'd expect it to, this branch can actually make most workflows a lot faster, even without concurrency.

### Notes for reviewer

For the time being, I'd like to focus on functionality, and making sure everything is feature complete, and all of the integration tests pass. There's probably some missing doc's, and I'll get to those things later.

Also, to make review easier, instead of making changes to `stacker.plan` I made a new `stacker.plan2`, and swapped out call sites to use the new graph based plan implementation. This will likely replace `stacker.plan` before merging, unless we want to keep the old `stacker.plan` around for API compatibility.

The `stacker.dag` module probably doesn't need much attention, since it's already been reviewed previously, and is stilled based on `py-dag`. Feel free to review it if you want.

### TODO

* [x] Bring back the check pointed loop logger (I just changed this to a simplified logger to make refactoring easier)
* [x] Bring back some removed tests.
* [x] Re-implement the `--tail` flag.
* [x] Re-implement dump

### CODE REVIEW

File: stacker/dag/__init__.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+class DAG(object):
+    """ Directed acyclic graph implementation. """
+
+    def __init__(self):
+        """ Construct a new DAG with no nodes or edges. """
+        self.reset_graph()
+
+    def add_node(self, node_name, graph=None):
+        """ Add a node if it does not exist yet, or error out. """
+        if not graph:
```
Comment (phobologic) [inline stacker/dag/__init__.py]: Do you have any idea why this library does this so often?  I haven't seen this pattern used really ever before.  I thought maybe for testing, but it seems like a messy way to handle that.

File: stacker/dag/__init__.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        """ Add an edge (dependency) between the specified nodes. """
+        if not graph:
+            graph = self.graph
+        if ind_node not in graph:
+            raise KeyError('node %s does not exist' % ind_node)
+        if dep_node not in graph:
+            raise KeyError('node %s does not exist' % dep_node)
+        test_graph = deepcopy(graph)
+        test_graph[ind_node].add(dep_node)
+        is_valid, message = self.validate(test_graph)
```
Comment (phobologic) [inline stacker/dag/__init__.py]: It seems like the `validate` method should throw the exception itself, and have anything that calls it deal with the results if they want to catch them.

File: stacker/dag/__init__.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+            graph[ind_node].add(dep_node)
+        else:
+            raise DAGValidationError(message)
+
+    def delete_edge(self, ind_node, dep_node, graph=None):
+        """ Delete an edge from the graph. """
+        if not graph:
+            graph = self.graph
+        if dep_node not in graph.get(ind_node, []):
+            raise KeyError('this edge does not exist in graph')
```
Comment (phobologic) [inline stacker/dag/__init__.py]: Can you add more info to this exception?  I actually think his overuse of KeyError, instead of custom Exceptions, is a little sloppy.  In any case, I think it's going to be necessary to at least have the two nodes in the exception, so we have a useful error message.

File: stacker/dag/__init__.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+                    edges.remove(old_task_name)
+                    edges.add(new_task_name)
+
+    def predecessors(self, node, graph=None):
+        """ Returns a list of all predecessors of the given node """
+        if graph is None:
+            graph = self.graph
+        return [key for key in graph if node in graph[key]]
+
+    def downstream(self, node, graph=None):
```
Comment (phobologic) [inline stacker/dag/__init__.py]: `descendants` maybe to match `predecessors` more?

When I was cleaning this up I called them: `ancestors` (recursive parents) and `descendants` (recursive children) - I also had `children` and `parents`, as well as properties for `has_children`, `has_descendants`, etc.  `has_children` and `has_descendants` were both the same call, because you couldn't have descendants if you didn't have children.

Comment (phobologic) [reply stacker/dag/__init__.py]: Oh, sorry, this is what I woul dhave called `children` - `all_downstream` is what I would have called `descendants`

File: stacker/dag/__init__.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+        # Add only the nodes we need.
+        for node in nodes:
+            dag.add_node_if_not_exists(node)
+            for edge in self.all_downstreams(node, graph=graph):
+                dag.add_node_if_not_exists(edge)
+
+        # Now, rebuild the graph for each node that's present.
+        for node, edges in graph.items():
+            if node in dag.graph:
```
Comment (phobologic) [inline stacker/dag/__init__.py]: This if seems dangerous, since we silently throw away any nodes that aren't in dag.graph, which really shouldn't ever happen, right?

File: stacker/dag/__init__.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+        """ Reset the graph and build it from the passed dictionary.
+
+        The dictionary takes the form of {node_name: [directed edges]}
+        """
+
+        self.reset_graph()
+        for new_node in graph_dict.iterkeys():
+            self.add_node(new_node)
+        for ind_node, dep_nodes in graph_dict.iteritems():
+            if not isinstance(dep_nodes, list):
```
Comment (phobologic) [inline stacker/dag/__init__.py]: I think you want to compare this against `collections.Iterable`, rather than just list.  A tuple would be perfectly ok  here as well.

File: stacker/dag/__init__.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+                raise TypeError('dict values must be lists')
```
Comment (phobologic) [inline stacker/dag/__init__.py]: Give the node name here, just so you can find the offending node in the case of an error.

File: stacker/dag/__init__.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+            for dep_node in dep_nodes:
+                self.add_edge(ind_node, dep_node)
+
+    def reset_graph(self):
+        """ Restore the graph to an empty state. """
+        self.graph = OrderedDict()
+
+    def ind_nodes(self, graph=None):
```
Comment (phobologic) [inline stacker/dag/__init__.py]: I think this is short for `independent_nodes` - can we name it that instead?

File: stacker/dag/__init__.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+            self.add_node(new_node)
+        for ind_node, dep_nodes in graph_dict.iteritems():
+            if not isinstance(dep_nodes, list):
+                raise TypeError('dict values must be lists')
+            for dep_node in dep_nodes:
+                self.add_edge(ind_node, dep_node)
+
+    def reset_graph(self):
+        """ Restore the graph to an empty state. """
+        self.graph = OrderedDict()
```
Comment (phobologic) [inline stacker/dag/__init__.py]: Is the fact that the graph is a dict going to cause issues with duplicate keys?  Like if someone puts two stack definitions in, could it result in the latter ovewriting the former?

We could still use OrderedDict, just sub-class it and override `__setitem__` so that it checks for duplicates if this is a concern.

File: stacker/dag/__init__.py [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+
+    def ind_nodes(self, graph=None):
+        """ Returns a list of all nodes in the graph with no dependencies. """
+        if graph is None:
+            graph = self.graph
+
+        dependent_nodes = set(
```
Comment (phobologic) [inline stacker/dag/__init__.py]: I'm having a really hard time reading this - any way to dumb it down that isn't going to destroy performance?

File: stacker/dag/__init__.py [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+                in_degree[v] -= 1
+                if in_degree[v] == 0:
+                    queue.appendleft(v)
+
+        if len(l) == len(graph):
+            return l
+        else:
+            raise ValueError('graph is not acyclic')
+
+    def size(self):
```
Comment (phobologic) [inline stacker/dag/__init__.py]: You could also give this class a `__len__` method so that calling `len()` on it would return the size.

File: stacker/exceptions.py [hunk 12]
```diff
@@ -217,3 +217,17 @@ def __init__(self, failed_stacks, *args, **kwargs):
...
+
+class GraphError(Exception):
+    """Raised when the graph is invalid (e.g. acyclic dependencies)
+    """
+
+    def __init__(self, exception, stack=None, dependency=None):
+        self.stack = stack
+        self.dependency = dependency
+        self.exception = exception
+        message = ("Error detected when adding '%s' "
```
Comment (phobologic) [inline stacker/exceptions.py]: This message will be broken if stack  & dependency aren't set.

File: stacker/plan2.py [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+        graph.add_step(step)
+
+    for step in steps:
+        for dep in step.requires:
+            graph.connect(step, dep)
+
+    return graph
+
+
+class Graph(object):
```
Comment (phobologic) [inline stacker/plan2.py]: Curious if this should just be combined with the `Plan` class?

File: stacker/plan2.py [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+
+    @property
+    def step_names(self):
+        return [step.name for step in self.steps]
+
+    def keys(self):
+        return self.step_names
+
+
+class UnlimitedSemaphore(object):
```
Comment (phobologic) [inline stacker/plan2.py]: Is this just planning ahead for multi-process/threading?

Comment (ejholmes) [reply stacker/plan2.py]: Yeah, this is leftover from multi-threading. I'll remove this since it's unnecessary right now.

File: stacker/plan2.py (hunk 13 above)
Comment (ejholmes) [reply stacker/plan2.py]: I originally started with that, but I think it's cleaner to keep these things separate:

1. `Graph` encapsulates a graph of `Step` objects, and has operations to make working on the graph itself easy (transposition, walking, etc).
2. `Plan` is more like a convenience class for Actions to consume (dump, outline, etc).

`Plan` could probably eventually get removed, as most of the methods on it may be better somewhere else, but I primarily kept it around to keep API compatibility with the existing `Plan` implementation.

**Comment by ejholmes** [comment]
Functional tests are all passing, so the only next steps to get this mergeable should be:

1. Implement `--tail` flag.
2. Implement `--dump` flag.
3. Re-implement the loop logger.

**Comment by ejholmes** [comment]
@phobologic when you wanna review this, this should be 100% feature complete and pretty close to being mergeable.

A lot of your original comments were around the implementation of `py-dag`. I'm not really inclined to change it very much, since it's algorithmically correct and has good test coverage. Would you rather pull in `py-dag` as a third party dependency, or leave it vendored as is?

File: stacker/exceptions.py (hunk 12 above)
Comment (ejholmes) [reply stacker/exceptions.py]: Where this is used, those things are never not present.

Comment (phobologic) [reply stacker/exceptions.py]: Cool - you should make them required then, so we don't hit an edge case.

**Comment by phobologic** [comment]
A few of my comments are around things that look dangerous or error prone (duplicate key comment, the comment about ignoring mising edges)- if you think they aren't and can explain it, I'd be fine ignoring those.

Others are just from general usefulness.  Better errors, easier to read code, etc.

I'm fine with skipping the stuff about children/descendants, since I know that's mostly just graph theory terms being used instead.

All in all, I'll take a small performance hit if it means better maintainability in the future.  That said, I don't expect my suggestions/questions to actually result in much of a hit at the scale we're operating and due to the fact that most of the time is spent in talking with AWS.

I think we should go ahead and vendor - though we should give the Author credit at the top of the code most likely.

Ok, now going to review the latest.

File: stacker/dag/__init__.py (hunk 1 above)
Comment (ejholmes) [reply stacker/dag/__init__.py]: No clue. It seems unnecessary. I'll try to add a commit that removes this.

File: stacker/tests/test_plan2.py [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+        app = Stack(
+            definition=generate_definition(
+                'app', 1, requires=['namespace-db.1']),
+            context=self.context)
+
+        with self.assertRaises(GraphError) as expected:
+            build_plan(
+                description="Test",
+                steps=[Step(vpc, None), Step(db, None), Step(app, None)])
+        message = ("Error detected when adding 'namespace-db.1' "
```
Comment (phobologic) [inline stacker/tests/test_plan2.py]: You know, you probably wouldn't need this message test if we had different exceptions ;)  Just saying.

File: stacker/__init__.py [hunk 16]
```diff
@@ -1 +1,25 @@
+from stacker.plan2 import Step, build_plan
+
 __version__ = "1.1.4"
+
+
+def plan(description=None, action=None,
```
Comment (phobologic) [inline stacker/__init__.py]: What's the plan going forward with this? We probably can't name this function "plan" if we move `plan2` to plan.

File: stacker/dag/__init__.py (hunk 1 above)
Comment (phobologic) [reply stacker/dag/__init__.py]: Thanks, yeah, I couldn't figure it out - I thought it'd be a test thing, but just now looking at tests I don't think it is.  Strange.

File: stacker/dag/__init__.py (hunk 5 above)
Comment (ejholmes) [reply stacker/dag/__init__.py]: That's what should happen. It's iterating through the full graph of nodes, and only re-adding edges for nodes that exist in the filtered graph.

Comment (phobologic) [reply stacker/dag/__init__.py]: Ahh, my bad - mixing dag & graph in this method made it confusing.  Maybe having it be `original_graph` and `filtered_graph` would make it less confusing.  Not a huge deal for now.

File: stacker/dag/__init__.py (hunk 9 above)
Comment (ejholmes) [reply stacker/dag/__init__.py]: That shouldn't be possible, since there's two places that check this now:

1. [Config level](https://github.com/remind101/stacker/blob/f07e6f2c6280d45d67e76a7d53434232b4312390/stacker/config/__init__.py#L370-L379)
2. [Graph level](https://github.com/remind101/stacker/blob/f9ba21857a2b7a17d4d58e2a6ab820da9e982fe9/stacker/dag/__init__.py#L28-L29)

I can add a functional test for this though, if we don't already have one.

File: stacker/__init__.py (hunk 16 above)
Comment (ejholmes) [reply stacker/__init__.py]: The plan (no pun intended) would be to replace `plan.py` with `plan2.py` before merging, unless we want to keep the original `plan.py` around for backwards compatibility. I only went with `plan2.py` because overwriting `plan.py` in the PR would make reviewing it pretty hard, since it's a re-write.

File: stacker/tests/test_plan2.py [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+        vpc = Stack(
+            definition=generate_definition('vpc', 1),
+            context=self.context)
+        bastion = Stack(
+            definition=generate_definition('bastion', 1, requires=[vpc.fqn]),
+            context=self.context)
+
+        plan = build_plan(description="Test", steps=[Step(vpc), Step(bastion)])
+
+        self.assertEqual(plan.graph.dag.graph, {
```
[the remark this answers is not in the document]
Comment (ejholmes) [reply stacker/tests/test_plan2.py]: This is really just a bad test on my part. `dag` should be private within the Graph class, but no way to enforce that with python :(.

I'll add a `to_dict` method to Graph, to make testing this cleaner.

File: stacker/tests/actions/test_destroy.py [hunk 18]
```diff
@@ -72,95 +86,3 @@ def test_destroy_stack_complete_if_state_submitted(self):
         # if we have processed the step and then can't find the stack, it means
         # we successfully deleted it
         self.assertEqual(status, COMPLETE)
-
-    def test_destroy_stack_in_parallel(self):
```
[the remark this answers is not in the document]
Comment (ejholmes) [reply stacker/tests/actions/test_destroy.py]: I don't think so. I think the purpose of the original test here was to test that the logic for ordering was correct, but that's now done at a lower level within the graph.

I should probably bring back the test below this one though (`test_destroy_stack_step_statuses`).

File: stacker/__init__.py (hunk 16 above)
Comment (phobologic) [reply stacker/__init__.py]: Right, but what about the `plan` method?  If you have a `plan` method in `stacker/__init__.py` and a `stacker/plan` directory, you're going to have conflicts when someone does something like: `from stacker import plan` or `from stacker.plan import *` (maybe not on the last one, I honestly don't know!)

File: stacker/tests/test_plan2.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (phobologic) [reply stacker/tests/test_plan2.py]: Cool - I also just realized the name of this test is probably wrong.

File: stacker/plan2.py [hunk 19]
```diff
@@ -0,0 +1,345 @@
...
+        self.set_status(SKIPPED)
+
+    def submit(self):
+        """A shortcut for set_status(SUBMITTED)"""
+        self.set_status(SUBMITTED)
+
+
+def build_plan(description=None, steps=None,
+               step_names=None, reverse=False):
+    graph = build_graph(steps)
```
[the remark this answers is not in the document]
Comment (ejholmes) [reply stacker/plan2.py]: I can't think of a scenario where that could actually happen from the `stacker` CLI, since there's config validation to ensure that stacks isn't empty, but I'll add a test for that here.

File: stacker/__init__.py (hunk 16 above)
Comment (ejholmes) [reply stacker/__init__.py]: Ah, gotcha. I think this `plan` method should actually be moved to `stacker.actions.base`, since that's where it's intended to be used.

File: tests/suite.bats [hunk 20]
```diff
@@ -629,10 +629,7 @@ EOF
   assert "$status" -eq 1
```
[the remark this answers is not in the document]
Comment (ejholmes) [reply tests/suite.bats]: In the implementation in this PR, if a node in the graph fails (returns False), the entire graph walk will halt. It's not actually necessary to halt the entire graph walk, only nodes that depend on the failed node. I can update the implementation of the graph walk to support the old behavior.

**Comment by ejholmes** [comment]
Alright @phobologic. I think I've addressed a majority of the comments here.

I also reverted back to the original behavior for failed child stacks, so that the entire graph of stacks is executed, but a given stack will not execute if any of it's dependencies are in a "not ok" state (see cd0c0f6b344444bee45992c1a166769b41c6b660)

**Comment by ejholmes** [comment]
@Lowercases found a bug where `CancelExecution` wasn't handled (pressing `n` in interactive mode). Added a commit that handles this flow, and also an integration test, since we didn't have one.

File: stacker/plan2.py [hunk 21]
```diff
@@ -0,0 +1,345 @@
...
+
+        return self.graph.walk(walk_func)
+
+    def execute(self, **kwargs):
+        return self.walk(**kwargs)
+
+    def walk(self):
+        """Walks each step in the underlying graph, in topological order."""
+
+        def walk_func(step):
```
[the remark this answers is not in the document]
Comment (ejholmes) [reply stacker/plan2.py]: I think we definitely should make that change, since it's annoying when a stack throws an exception, and you have no context about what stack has the issue. But we should probably make that change separate from this PR, since that's not existing behavior.

We'll also need to think about how we handle exceptions in a parallelized world, since the current method of just throwing an exception and aborting won't work with multi-threads/processes.

File: stacker/plan2.py [hunk 22]
```diff
@@ -0,0 +1,400 @@
...
+
+    for step in steps:
+        for dep in step.requires:
+            graph.connect(step, dep)
+
+    return graph
+
+
+class Graph(object):
+    """Graph represents a graph of steps.
```
Comment (phobologic) [inline stacker/plan2.py]: This is super helpful - thanks!

**Comment by ejholmes** [comment]
Cool. I'll update us internally to run this branch before merging.

**Comment by ejholmes** [comment]
There's some talk about whether or not the `--stacks` flag should stop at the given stack, or continue to execute stacks that transitively depend on the provided stack.

Let's say you have a graph like the following:



In the current behavior, if you specify `--stacks db`, stacker will effectively filter the graph to the following:



If there's a change to the `db` stack (like a new instance url or w/e), then the `app` stack won't see it.

We could change the behavior here so that the graph includes the dependencies of `db`, as well as anything that transitively depends on it. So the graph would be filtered to:

**Comment by acmcelwee** [comment]
Personally, the expected behavior of the `--stacks` flag is as you described it in the first example, and how it's currently implemented in the PR. That said, I think the second behavior could be useful in other occasions. It might be worth thinking about how both could coexist with different flags or modes.

**Comment by russellballestrini** [comment]
The version is this PR is `faster` but I'm pretty sure the version in the 3rd chart is `safer` (and puts the system as a whole in a more correct state).

I would prefer correctness over speed.

I like @acmcelwee point, if we can support both, (`--stacks-fast`) I think that could be really cool.

**Comment by ejholmes** [comment]
Yeah, that makes sense to me. I think I'll update this PR to the more complete version.

If there's a need for the old behavior, we can add new flags for that.

**Comment by ejholmes** [comment]
Aaaaand...I think I'm gonna flipflop my preference back to the fast version. I tried the complete version internally, and it's pretty easy for the graph to explode making `--stacks` mostly useless.

I'll leave out specific stack names, but on our internal graph, when I targeted a stack that builds out an ECS cluster, the fast versions results in 4 stacks (3 dependencies). With the full version, the graph explodes to 100 stacks as a result of the parents transitive dependencies.

**Comment by ejholmes** [comment]
Thanks everyone for the comments and feedback. Super excited about what this is gonna enable for stacker moving forward.
